rfcs/004 - Better Policy UX RFC.md

```diff
@@ -0,0 +1,156 @@
+# 004 - Better Policy UX RFC
+
+## Current Status
+
+### Proposed
+
+2022-09-06
+
+### Accepted
+
+YYYY-MM-DD
+
+#### Approvers
+
+- Full Name <email@example.com>
+
+### Implementation
+
+- [ ] Implemented: [One or more PRs](https://github.com/iasql/iasql-engine/some-pr-link-here) YYYY-MM-DD
+- [ ] Revoked/Superceded by: [RFC ###](./000 - RFC Template.md) YYYY-MM-DD
+
+## Author(s)
+
+- Joseph <joseph@iasql.com>
+
+## Summary
+
+Customers (even the very technically experienced ones) often have to look up how to construct an AWS IAM policy and then spend time validating the policy before inserting into IaSQL.
+This becomes more and more unmanageable as the policy grows.
+This is not a good experience, and it can be improved upon. We believe we can collect some inputs from IaSQL customers and offer then some form of automation when creating policies, so they can do it faster, more confidently and more conveniently.
+
+## Proposal
+
+We should add new `Policy` and `Statement` entities (perhaps in the `aws_iam` module). These entities would serve as 
+a way to construct IAM policies in a more user-friendly way. For example, the customer intends to create an S3 bucket with
+read-only permissions from all Principals within the same AWS account. Ideally, something like this:
+
+```json
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Effect": "Allow",
+      "Resource": "aws:arn:123456789012:bucket-resource",
+      "Action": [
+        "s3:GetObjects"
+      ],
+      "Principal": {
+        "AWS": "arn:aws:iam::123456789012:root"
+      }
+    }
+  ]
+}
+```
+
+Instead of the customer having to look up and put together a JSON blob like this and then verifying and then inserting, we can also give them
+the option to do this using functions that would generate this blob.
+
+```sql
+-- User creates policy
+select * from iasql_create_policy(
+    'some_policy_name',
+    '2012-10-17' -- nullable, defaults to 2012-10-17
+    );
+
+-- User creates statement and attaches to the policy
+select * from iasql_create_policy_statement(
+        '<policy_id>',
+        'allow',
+        'aws:arn:123456789012:bucket-resource',
+        's3:GetObjects', -- comma separated list of actions
```

Comment:
The actions still being strings means that if I typo `s3:GetObjects` as `s3;GetObjects` it will fail on me and IaSQL will provide no warning ahead of time that this is the case.

Reply:
I have addressed this in the second draft by suggesting regex checks and possibly a call to the AWS SDK to validate the policy